server.go

```diff
@@ -4173,13 +4327,297 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+		// do anything.
+		return nil
+
+	case temporary:
+		// If this peer's access status is temporary, we'll need to
+		// update the pendingChannelPeers map. The peer's access status
+		// will stay temporary.
+		numPending, pendingFound := s.pendingChannelPeers[remoteHex]
+		if !pendingFound {
+			// Error if we did not find any info in
+			// pendingChannelPeers.
+			return ErrNoPendingPeerInfo
+		}
+
+		// Increment the numPending amount.
+		s.pendingChannelPeers[remoteHex] = numPending + 1
+
+	case restricted:
+		// If the peer's access status is restricted, then we can
+		// transition it to a temporary-access peer. We'll need to
+		// update numRestricted and also peerScores. We'll also need to
+		// update pendingChannelPeers.
+		s.pendingChannelPeers[remoteHex] = 1
+
+		// A restricted-access slot has opened up.
+		s.numRestricted -= 1
+
+		s.peerScores[remoteHex] = peerSlotStatus{state: temporary}
+
+	default:
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+
+	return nil
+}
+
+// NewPendingCloseChan is called when a pending-open channel prematurely closes
+// before the funding transaction has confirmed. This potentially demotes a
+// temporary-access peer to a restricted-access peer. If no restricted-access
+// slots are available, the peer will be disconnected.
+func (s *server) NewPendingCloseChan(remoteHex string) error {
+	// Ensure that we don't lock banScoreMtx and _then_ lock the server
+	// mutex. This could lead to deadlock.
+	s.banScoreMtx.Lock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		s.banScoreMtx.Unlock()
+
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case protected:
+		s.banScoreMtx.Unlock()
+
+		// If this peer is protected, we don't do anything.
+		return nil
+
+	case temporary:
+		// If this peer is temporary, we need to check if it will revert
+		// to a restricted-access peer.
+		numPending, pendingFound := s.pendingChannelPeers[remoteHex]
+		if !pendingFound {
+			s.banScoreMtx.Unlock()
+
+			// Error if we did not find any info in
+			// pendingChannelPeers.
+			return ErrNoPendingPeerInfo
+		}
+
+		currentNumPending := numPending - 1
+
+		if currentNumPending == 0 {
+			// Remove the entry from pendingChannelPeers.
+			delete(s.pendingChannelPeers, remoteHex)
+
+			// If this is the only pending-open channel for this
+			// peer and it's getting removed, attempt to demote
+			// this peer to a restricted peer.
+			if s.numRestricted == maxRestrictedSlots {
+				// There are no available restricted slots, so
+				// we need to disconnect this peer.
+				s.banScoreMtx.Unlock()
+
+				remotePub, err := btcec.ParsePubKey(
+					[]byte(remoteHex),
+				)
+				if err != nil {
+					return err
+				}
+
+				// The banScoreMtx cannot be held when the
+				// below DisconnectPeer call occurs.
+				return s.DisconnectPeer(remotePub)
```

Comment:
iiuc we could end up demoting more than one peer at the same time.

Reply:
How so?

Reply:
IIUC when we decrement `s.numRestricted` when we disconnect and then remove the peer. But the `banScoreMtx` is lifted in between so it could be that if we reach the ceiling then concurrent calls to this function would still see the same counts. Doesn't seem to be an issue to me, more just a remark.

Reply:
unit-tests for those function would be very nice